docs: debug commands, lifecycle steps, helm healing and deduplication (#87)
flant/addon-operator: LIFECYCLE.md

The module files are located in the `/modules` directory. The directory can be set via $MODULES_DIR variable. The global hook files are located in the `/global-hooks` directory (you can set your own directory with the $GLOBAL_HOOKS_DIR variable).

# Lifecycle

> ## startup sequence

Reviewer: All headings should be capitalized.

Author: Done

During the start, Addon-operator finds and initializes all global hooks. For more info, see [HOOKS](HOOKS.md#initialization-of-global-hooks).

After the global hooks initialization, Addon-operator executes all global `onStartup` hooks.

Then, the *global* hooks with `kubernetes` binding are executed with a binding context of type `Synchronization` and Kubernetes monitors for global hooks are started.

## reload all modules

> Next, the 'reload all modules' process is started. It firstly finds all modules and their hooks.

Reviewer: ```suggestion
Next, the 'reload all modules' process is started. First, it finds all modules and their hooks.
```

Then, all global hooks with `beforeAll` binding are executed.
 
Next, the 'module discovery' process is started, it finds which modules are enabled and starts them.

> During each module start-up, it executes all `onStartup` hooks and initializes the installation of a helm chart. Prior to the installation of a helm chart, the `beforeHelm` hook is executed. The `afterHelm` hook is executed after the installation.

Reviewer: ```suggestion
During each module start-up, it executes all `onStartup` hooks and initializes the installation of a Helm chart. Prior to the installation of a Helm chart, the `beforeHelm` hook is executed. The `afterHelm` hook is executed after the installation.
```

When all modules are started, all global hooks with `afterAll` binding are executed.

## main loop

After the first run of 'reload all modules', the main loop starts. It reacts to schedule and Kubernetes events, and to a values changes: it restarts a particular module if its values are changed and runs 'reload all modules' process again if global values are changed.

## named queues

> The Addon-operator supports named queues to execute hooks in parallel for bindings `schedule` and `kubernetes/Event`.

Reviewer: ```suggestion
The Addon-operator supports named queues to execute hooks in parallel for `schedule` and `kubernetes/Event` bindings.
```

> All other actions are handled in one "main" queue:

Reviewer: ```suggestion
All other actions are handled in a single "main" queue:
```

- global hooks:
  - `onStartup`
  - `kubernetes/Synchronization`
  - `beforeAll`
  - `afterAll`
- module hooks:
  - `onStartup`
  - `kubernetes/Synchronization`
  - `beforeHelm`

>   - executions of `helm` commands

Reviewer: ```suggestion
  - execution of `helm` commands
```

  - `afterHelm`

> A detailed picture of the lifecycle is available in the [LIFECYCLE-STEPS](LIFECYCLE-STEPS.md) document.

Reviewer: ```suggestion
A detailed view of the lifecycle process is available in the [LIFECYCLE-STEPS](LIFECYCLE-STEPS.md) document.
```

Author: ```suggestion
This document mainly describes modules. To get more information on hooks, see [HOOKS](HOOKS.md) document. To get a full view of how hooks, modules, [values](VALUES.md), binding contexts, and queues are interlinked, see [LIFECYCLE-STEPS](LIFECYCLE-STEPS.md) document.
```

# Module lifecycle

The `onStartup` hooks of all modules are executed during the first deployment of a Pod with an Addon-operator.

> Next, the modules are run in alphabetical order with `helm upgrade --install`. Prior to launching helm, `beforeHelm` hooks are executed, after the launch, the `afterHelm` hooks are executed.

Reviewer: ```suggestion
Next, the modules are run in alphabetical order with `helm upgrade --install`. Prior to launching Helm, `beforeHelm` hooks are executed, after the launch, the `afterHelm` hooks are executed.
```

After the launch module would start to respond to two types of events:

- `schedule` — events that are generated by the crontab scheduler built in the addon-operator;
- `kubernetes` — events within the cluster that api-server announces to Addon-operator.

> When the module is deactivated, the Addon-operator launch command `helm delete --purge` and after the release deletion, the `afterDeleteHelm` hooks are executed.

Reviewer: ```suggestion
When the module is deactivated, the Addon-operator launches command `helm delete --purge` and after the release deletion, the `afterDeleteHelm` hooks are executed.
```

Module `some-module` is explicitly disabled in `modules/values.yaml` but enabled by `someModuleEnabled` key in ConfigMap/addon-operator. Thus enabled script is executed and returns `false`. So the final result is that the module is disabled.

# Task queues

> Task queues are a simple FIFO queues. The Addon-operator process an event, creates a task and adds it to the particular named queue. Each named queue has its own queue handler which runs the top task and proceeds to the next.

Reviewer: "Top" is suitable for a stack, not for a queue.
```suggestion
Task queues are simple FIFO queues. The Addon-operator processes an event, creates a task and adds it to the particular named queue. Each named queue has its own queue handler which runs the first task and proceeds to the next.
```

> Each task is processed until successful completion. In case of an error, the task is returned to the top of the queue and executed after a 5 seconds delay. When executing tasks for the `kubernetes` and `schedule` events, the queue handler ignores the execution errors if the `allowFailure: true` flag is specified in the binding configuration.

Reviewer: ```suggestion
Each task is processed until successful completion. In case of an error, the task is returned to the start of the queue and executed after a 5 seconds delay. When executing tasks for the `kubernetes` and `schedule` events, the queue handler ignores the execution errors if the `allowFailure: true` flag is specified in the binding configuration.
```
